### ISSUE TITLE
[await-thenable] Unexpected `await` of a non-Promise does not work if wrapped with Promise.all

### ISSUE THREAD

**Issue by viestat** [typescript-eslint/typescript-eslint#1804]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/await-thenable": "error"
  }
}
```

```TS
const value = 'value';
const createValue = () => 'value';

await Promise.all([value, createValue]);
```

**Expected Result**
The rule should show a linting error:
```
Unexpected `await` of a non-Promise (non-"Thenable") value  @typescript-eslint/await-thenable
```

**Actual Result**
No liniting error.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.24.0` |
| `@typescript-eslint/parser`        | `2.24.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.10.0` |
| `npm`                              | `6.10.3` |

**Comment by bradzacher** [comment]
I am unable to reproduce this against master

![image](https://user-images.githubusercontent.com/7462525/77668447-a7add980-6f40-11ea-9cee-3873441aeb10.png)


Could you please provide a repro repo for this to help me investigate?

**Comment by viestat** [comment]
@bradzacher I will create the repro repo, In the meantime, from the image you shared I can see that you added this test case in the **valid** property,  the issue is that the code should be **invalid** and throw an error.

**Comment by bradzacher** [comment]
wait sorry, I misunderstood your OP.

Why should it be invalid?
`Promise.all` returns a `Promise` always, regardless of the input values:
![image](https://user-images.githubusercontent.com/7462525/77673281-345b9600-6f47-11ea-989e-79aa21f38cce.png)
Which means that it should always be awaited.

**Comment by viestat** [comment]
> Why should it be invalid?
> `Promise.all` returns a `Promise` always, regardless of the input values 

You are right,  While it is valid JavaScript to pass non-promise elements to `Promise.all` it seems to be inconsistent with this rule.
It would prevent code like the following to be written:
```TS
const promise1 = Promise.resolve();
const promise2 = Promise.resolve();
const test1 = (): void => {};

export const run = async (): Promise<void> => {
  await Promise.all([promise1, promise2, test1()]);
};
```
Here it makes no sense at all to use `await` for all the input values since one of them is not a promise.

**Comment by bradzacher** [comment]
AHHHHHH okay i'm with you now. Sorry, slow morning.

You're saying that you want the rule to warn against passing non-promise values into the native promise aggregation functions (`Promise.all`, `Promise.allSettled`, `Promise.race`).

If yes, this sounds perfectly reasonable. Would be good as an option on the rule.

**Comment by viestat** [comment]
@bradzacher Precisely 😄

**Comment by andreialecu** [comment]
Refactored some code recently where I saw various calls like:

```
const someFn = () => {...};
await Promise.all([await someFn(), await someOtherFn()])
```

It seems this should also prevent that type of problem? Or is there an alternative?

**Comment by bradzacher** [comment]
No, there are no rules to prevent this code. Please file a new issue, and happy to accept a PR

**Comment by andreialecu** [comment]
I understand that, I meant that fixing issue would additionally fix the use case I mentioned.

**Comment by bradzacher** [comment]
Sorry, I guess that potentially it would be if the types work out (i.e. your functions don't return something weird like `Promise<Promise<foo>>`.

Though it wouldn't be warning against the fact that you have `await foo` in the `Promise.all`, it would instead warn against passing a non-promise into `Promise.all` - meaning the error messages will be relatively cryptic.

**Comment by andreialecu** [comment]
I think that if functions return `Promise<Promise<foo>>` then the usage is probably correct and there shouldn't be a warning. 

I think the cryptic error would be ok, at least the user will get a clue that something is wrong with the code. Perhaps it could even provide a hint like `did you already await the inner promise?`

**Comment by hazae41** [comment]
Day since last bug caused by this: 0

**Comment by samchungy** [comment]
We ran into a prod incident because of this.

This is a dumbed down version of what happened:

```ts
await Promise.all([asyncGetData(), asyncGetData2(), syncGetData()]);
```

`syncGetData` threw a error which meant that `asyncGetData` and `asyncGetData2` were left as dangling promises which meant we ran into unhandled promise exceptions when `asyncGetData` or `asyncGetData2` failed

EDIT:

To illustrate what's happening this would be the equivalent:

```ts
const promise1 = asyncGetData();
const promise2 = asyncGetData2();
syncGetData(); // throws error so promise1 and promise2 are left dangling
await Promise.all([promise1, promise2, notPromise]);
```

**Comment by kirkwaiblinger** [comment]
@samchungy Fun fact, even with this check (which is close to landing, see #11267), you won't be _fully_ safe from synchronous errors, since promise-returning functions can still throw synchronously 🙃 

```ts
async function asyncGetData1() {
  if (Math.random() > 0.5) throw new Error('rejected promise');
  return 'data';
}

function asyncGetData2() {
  if (Math.random() > 0.5) throw new Error('threw synchronously');
  return Promise.resolve('data');
}

await Promise.all([asyncGetData1(), asyncGetData2()]);
```

**Comment by samchungy** [comment]
Aha yep I'm aware of that 😩, was just a bummer to run into. 

Something like: 

```ts
await Promise.all([asyncGetData(), asyncGetData(), asyncGetData(syncGetData());
```

Would also cause issues
